storage/operations.go

```diff
@@ -0,0 +1,139 @@
+package storage
+
+import (
+	"io"
+)
+
+// Iterator is an interface for iterating over key-value pairs in a storage backend.
+type Iterator interface {
+	// SeekGE seeks to the smallest key greater than or equal to the given key.
+	SeekGE()
+
+	// Valid returns whether the iterator is positioned at a valid key-value pair.
+	Valid() bool
+
+	// Next advances the iterator to the next key-value pair.
+	Next()
+
+	// Key returns the key of the current key-value pair, or nil if done.
+	IterItem() IterItem
+
+	// Close closes the iterator. Iterator must be closed, otherwise it causes memory leak.
+	Close() error
+}
+
+// IterItem is an interface for iterating over key-value pairs in a storage backend.
+type IterItem interface {
+	Key() []byte
+
+	// Value returns the value of the current key-value pair
+	// The reason it takes a function is to follow badgerDB's API pattern
+	Value(func(val []byte) error) error
+}
+
+type IteratorOption struct {
+	IterateKeyOnly bool // default false
+}
+
+func DefaultIteratorOptions() IteratorOption {
+	return IteratorOption{
+		IterateKeyOnly: false, // only needed for badger. ignored by pebble
+	}
+}
+
+// Reader is an interface for reading from a storage backend.
+type Reader interface {
+	// Get gets the value for the given key. It returns ErrNotFound if the DB
+	// does not contain the key.
+	//
+	// The caller should not modify the contents of the returned slice, but it is
+	// safe to modify the contents of the argument after Get returns. The
+	// returned slice will remain valid until the returned Closer is closed. On
+	// success, the caller MUST call closer.Close() or a memory leak will occur.
+	Get(key []byte) (value []byte, closer io.Closer, err error)
```

Comment:
The comments is taken from pebble's document, which says "The returned slice will remain valid until the returned Closer is closed". This gives me impression that once `closer.Close()` is called, then the value []byte is no longer available, which means can't be read from it anymore. 

But I just tested to always call closer.Close() before reading the value, seems OK that value can still be read and all test cases passed.

Reply:
Sorry, I didn't realize that - in that case I think we should leave it as you have originally implemented it, even if it appears to work in test cases.

Most likely Pebble is "freeing" that slice of memory to be re-used at some point in the future when `Close` is called. Even if in a test case, it remains valid for some time after calling `Close`, I would trust their documentation and assume that at some point it will be re-used, which could easily cause subtle bugs in our logic.